### PR TITLE
Remove GenericFlow && Adapt GET flow/stats/<dpid>

### PR DESCRIPTION
**Pull request by gretelliz**
This is related to issues #30 #35 #39 #40 #41, and PR #38.

-- The ``GenericFlow abstraction`` has been removed and the dict `self.flows_stats_dict` of `napps.kytos.of_core.v0x04.flow.Flow` instances is keept in memory by `flow.id`:

Example:
```
self.flows_stats_dict = {
    '6ea12166476c089c7fb8c868761032d2': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e44bb0>,
    '8c62f30f8fd0a15a6d0051d77555c246': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e44190
}
```
New flows from kytos/of_core.flow_stats.received are added to the dict. Should it be updated with the new flows and to remove what was previously in memory? ([see here](https://github.com/kytos-ng/flow_stats/blob/4120f52b01204f6718ea31eaa58f93982d1dffaf/main.py#L180))

-- `GET flow/match/<dpid>` endpoint was deleted.

-- I'm looking forward to knowing if endpoints `GET packet_count/sum/{{dpid}}` and `GET /bytes_count/sum/{{dpid}}` should be removed. This is related to issue #39 

-- `GET flow/stats/<dpid>` endpoint has been adapted. Return a dict of stats per flow given a list of switches.dpid.

Examples:

1. ``GET http://127.0.0.1:8181/api/amlight/flow_stats/flow/stats?dpid=00:00:00:00:00:00:00:01&dpid=00:00:00:00:00:00:00:02``
```
{
    "00:00:00:00:00:00:00:01": {
        "0d024b8d679624d6c87239f0cc2bb1c8": {
            "byte_count": 420,
            "duration_nsec": 384000000,
            "duration_sec": 115,
            "packet_count": 6
        },
        "5cd66b3ec5c0437b3d163f18e8491250": {  
            "byte_count": 339780,
            "duration_nsec": 26000000,
            "duration_sec": 118,
            "packet_count": 8090
        },
        "6ea12166476c089c7fb8c868761032d2": {  
            "byte_count": 444,
            "duration_nsec": 273000000,
            "duration_sec": 27893,
            "packet_count": 6
        },
        "8c62f30f8fd0a15a6d0051d77555c246": {  
            "byte_count": 74,
            "duration_nsec": 550000000,
            "duration_sec": 385,
            "packet_count": 1
        },
        "b15f4afad81af0dbd021dc871505109c": {  
            "byte_count": 80374,
            "duration_nsec": 267000000,
            "duration_sec": 27893,
            "packet_count": 191
        },
        "bd88376cf7c8069c42a28ed71b8867d5": {  
            "byte_count": 420,
            "duration_nsec": 770000000,
            "duration_sec": 90,
            "packet_count": 6
        }
    },
    "00:00:00:00:00:00:00:02": {
        "17273e132dd9e9c97962cd3dd8c8a7d8": {  
            "byte_count": 74,
            "duration_nsec": 947000000,
            "duration_sec": 394,
            "packet_count": 1
        },
        "23bfec0be8fabf35e1e74c2307b3b0a8": {  
            "byte_count": 679812,
            "duration_nsec": 810000000,
            "duration_sec": 126,
            "packet_count": 16186
        },
        "2e8155af1a85644a166b28deb5af1f79": {  
            "byte_count": 157906,
            "duration_nsec": 219000000,
            "duration_sec": 27893,
            "packet_count": 370
        },
        "9ba6573b12f2c0f0c8f98dbc1d366ad8": {  
            "byte_count": 0,
            "duration_nsec": 945000000,
            "duration_sec": 394,
            "packet_count": 0
        },
        "c7d11070198ebad1796b22c6d215c74b": {  
            "byte_count": 592,
            "duration_nsec": 231000000,
            "duration_sec": 27893,
            "packet_count": 8
        },
        "e56d326406a2c309f260c20bf6b8ece1": {  
            "byte_count": 444,
            "duration_nsec": 222000000,
            "duration_sec": 27893,
            "packet_count": 6
        }
    }
}
```

2. ``GET http://127.0.0.1:8181/api/amlight/flow_stats/flow/stats?dpid=00:00:00:00:00:00:00:04``
```
{
    "00:00:00:00:00:00:00:04": {
        "b6e13b9e2ab3f8bbcfff41d36021ec02": {  
            "byte_count": 342300,
            "duration_nsec": 69000000,
            "duration_sec": 301,
            "packet_count": 8150
        },
        "f833a9dee03fe10db9491282897965ce": {  
            "byte_count": 82213,
            "duration_nsec": 297000000,
            "duration_sec": 28073,
            "packet_count": 214
        }
    }
}
```

3. ``GET http://127.0.0.1:8181/api/amlight/flow_stats/flow/stats``
```
{
    "00:00:00:00:00:00:00:01": {
        "0d024b8d679624d6c87239f0cc2bb1c8": {
            "byte_count": 420,
            "duration_nsec": 988000000,
            "duration_sec": 59,
            "packet_count": 6
        },
        "5cd66b3ec5c0437b3d163f18e8491250": {  
            "byte_count": 343098,
            "duration_nsec": 73000000,
            "duration_sec": 358,
            "packet_count": 8169
        },
        "6ea12166476c089c7fb8c868761032d2": {  
            "byte_count": 444,
            "duration_nsec": 320000000,
            "duration_sec": 28133,
            "packet_count": 6
        },
        "8c62f30f8fd0a15a6d0051d77555c246": {  
            "byte_count": 74,
            "duration_nsec": 597000000,
            "duration_sec": 625,
            "packet_count": 1
        },
        "b15f4afad81af0dbd021dc871505109c": {  
            "byte_count": 80577,
            "duration_nsec": 314000000,
            "duration_sec": 28133,
            "packet_count": 192
        },
        "bd88376cf7c8069c42a28ed71b8867d5": {  
            "byte_count": 420,
            "duration_nsec": 989000000,
            "duration_sec": 119,
            "packet_count": 6
        }
    },
    "00:00:00:00:00:00:00:02": {
        "17273e132dd9e9c97962cd3dd8c8a7d8": {  
            "byte_count": 74,
            "duration_nsec": 960000000,
            "duration_sec": 574,
            "packet_count": 1
        },
        "23bfec0be8fabf35e1e74c2307b3b0a8": {  
            "byte_count": 686448,
            "duration_nsec": 858000000,
            "duration_sec": 366,
            "packet_count": 16344
        },
        "2e8155af1a85644a166b28deb5af1f79": {  
            "byte_count": 158312,
            "duration_nsec": 267000000,
            "duration_sec": 28133,
            "packet_count": 372
        },
        "9ba6573b12f2c0f0c8f98dbc1d366ad8": {  
            "byte_count": 0,
            "duration_nsec": 958000000,
            "duration_sec": 574,
            "packet_count": 0
        },
        "c7d11070198ebad1796b22c6d215c74b": {  
            "byte_count": 592,
            "duration_nsec": 279000000,
            "duration_sec": 28133,
            "packet_count": 8
        },
        "e56d326406a2c309f260c20bf6b8ece1": {  
            "byte_count": 444,
            "duration_nsec": 270000000,
            "duration_sec": 28133,
            "packet_count": 6
        }
    },
    "00:00:00:00:00:00:00:03": {
        "48083b8562161a2923d5edda794999f4": {  
            "byte_count": 686364,
            "duration_nsec": 847000000,
            "duration_sec": 363,
            "packet_count": 16342
        },
        "6055f13593fad45e0b4699f49d56b105": {  
            "byte_count": 160888,
            "duration_nsec": 223000000,
            "duration_sec": 28133,
            "packet_count": 381
        },
        "91b620cbd6760f59767b67f9c506a7a9": {
            "byte_count": 70,
            "duration_nsec": 931000000,
            "duration_sec": 3,
            "packet_count": 1
        },
        "94b2776a178f272ad4307fa810dcb2f4": {  
            "byte_count": 70,
            "duration_nsec": 994000000,
            "duration_sec": 119,
            "packet_count": 1
        },
        "b6f7b4c6e0a3817956c75eeec36c88a3": {  
            "byte_count": 592,
            "duration_nsec": 234000000,
            "duration_sec": 28133,
            "packet_count": 8
        },
        "e3d5e5426d30ad419e5eac107055c107": {  
            "byte_count": 0,
            "duration_nsec": 783000000,
            "duration_sec": 571,
            "packet_count": 0
        }
    },
    "00:00:00:00:00:00:00:04": {
        "b6e13b9e2ab3f8bbcfff41d36021ec02": {  
            "byte_count": 343140,
            "duration_nsec": 83000000,
            "duration_sec": 361,
            "packet_count": 8170
        },
        "f833a9dee03fe10db9491282897965ce": {  
            "byte_count": 82213,
            "duration_nsec": 311000000,
            "duration_sec": 28133,
            "packet_count": 214
        }
    }
}
```

These results are checked from the console:
```
kytos $> controller.napps[('amlight', 'flow_stats')].flows_stats_dict                                                                                                                                       
Out[2]: 
{'6ea12166476c089c7fb8c868761032d2': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e44bb0>,
 '8c62f30f8fd0a15a6d0051d77555c246': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e44190>,
 'bd88376cf7c8069c42a28ed71b8867d5': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e41640>,
 '5cd66b3ec5c0437b3d163f18e8491250': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e410d0>,
 'b15f4afad81af0dbd021dc871505109c': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e41580>,
 'b6e13b9e2ab3f8bbcfff41d36021ec02': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e3ea00>,
 'f833a9dee03fe10db9491282897965ce': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15f5760>,
 'b6f7b4c6e0a3817956c75eeec36c88a3': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15d8bb0>,
 'e3d5e5426d30ad419e5eac107055c107': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15f50a0>,
 '94b2776a178f272ad4307fa810dcb2f4': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15cbfa0>,
 '48083b8562161a2923d5edda794999f4': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15cb5e0>,
 '6055f13593fad45e0b4699f49d56b105': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c6688ca0>,
 'c7d11070198ebad1796b22c6d215c74b': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15e0790>,
 'e56d326406a2c309f260c20bf6b8ece1': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15e0970>,
 '17273e132dd9e9c97962cd3dd8c8a7d8': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15e0b50>,
 '9ba6573b12f2c0f0c8f98dbc1d366ad8': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15e0d30>,
 '23bfec0be8fabf35e1e74c2307b3b0a8': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c15e0eb0>,
 '2e8155af1a85644a166b28deb5af1f79': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84bedb20d0>,
 '0d024b8d679624d6c87239f0cc2bb1c8': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c66ea880>
 '91b620cbd6760f59767b67f9c506a7a9': <napps.kytos.of_core.v0x04.flow.Flow at 0x7f84c2e25c70>}

kytos $>  [f.switch.dpid for f in controller.napps[('amlight', 'flow_stats')].flows_stats_dict.values()]                                                                                                    
Out[3]: 
['00:00:00:00:00:00:00:01',
 '00:00:00:00:00:00:00:01',
 '00:00:00:00:00:00:00:01',
 '00:00:00:00:00:00:00:01',
 '00:00:00:00:00:00:00:01',
 '00:00:00:00:00:00:00:04',
 '00:00:00:00:00:00:00:04',
 '00:00:00:00:00:00:00:03',
 '00:00:00:00:00:00:00:03',
 '00:00:00:00:00:00:00:03',
 '00:00:00:00:00:00:00:03',
 '00:00:00:00:00:00:00:03',
 '00:00:00:00:00:00:00:02',
 '00:00:00:00:00:00:00:02',
 '00:00:00:00:00:00:00:02',
 '00:00:00:00:00:00:00:02',
 '00:00:00:00:00:00:00:02',
 '00:00:00:00:00:00:00:02',
 '00:00:00:00:00:00:00:01',
 '00:00:00:00:00:00:00:03']
```

-- `CHANGELOG.rst` and `openapi.yml` have been update

### CODE REVIEW

File: main.py
Comment (viniarck) [inline main.py:216]: @gretelliz you'll need to coordinate strategically when this PR will land, otherwise it'll break [`sdntrace_cp` ](https://github.com/kytos-ng/sdntrace_cp/blob/master/automate.py#L38), `sdntrace_cp` `Automate`'s related methods should also get the flows from `/flow_manager/v2/stored_flows`, so it'd be great to have a PR there to land these two shortly after both are approved.  

Also, `sdntrace_cp` used to store them in [`self._circuits`](https://github.com/kytos-ng/sdntrace_cp/blob/master/automate.py#L78), and it used to be updated on every event `amlight/flow_stats.flows_updated` that's being removed, so that can also be improved (removing that shared memory structure),  no need keep to subscribe to this event anymore, and then whenever it needs the flows it fetches from flow_manager as you've done and compute on-the-fly the transformations, since the scheduled methods won't run that frequent we shouldn't expect much surprises in terms of making too much requests.  Automated checks from sdntrace_cp are disabled by default, but we need to make sure it still works if in prod it's enabled.

File: openapi.yml
Comment (viniarck) [inline openapi.yml:7]: One of the requirements of EP32 stabilization was also to bump v1 on endpoint routes, could you update it accordingly since we're going for a stabilization?

**Review by viniarck — COMMENTED** [review]
@gretelliz, great to see how this is shaping up, and seeing generic_flows being removed and the API routes getting officially set for a v1. Thanks also for exploring the PR locally and providing the info, that was helpful for the review. I left some threads open. 

Could you also improve the code coverage, it has dropped 12%. If it could be at least as much as it used to that'd be great, and then you could also map a new issue to eventually try to cover 80% in a next opportunity.

**Comment by viniarck** [comment]
> New flows from kytos/of_core.flow_stats.received are added to the dict. Should it be updated with the new flows and to remove what was previously in memory? ([see here](https://github.com/kytos-ng/flow_stats/blob/4120f52b01204f6718ea31eaa58f93982d1dffaf/main.py#L180))

I think it's great as it is, even though they're not getting removed, if those flows are gone or accidentally removed, then operators would like to see that they're stats are not incrementing which would be great for monitoring, so, it's aligned I think. 

> -- I'm looking forward to knowing if endpoints GET packet_count/sum/{{dpid}} and GET /bytes_count/sum/{{dpid}} should be removed. This is related to issue https://github.com/kytos-ng/flow_stats/issues/39

In the EP032 blueprint we had an item stating that they'd be reassessed, during the review no one raised concerns there, please double check with @italovalcy and @jab1982 if they had any plans to use it in prod. That entails to what we're calling here the V1 stabilization of the API. Once we have a requirements to migrate this to the DB in the future, I think we could review these aggregations and provide them efficiently, so going for the removal here, while not adding more tech debt with shared memory structures sounds good to me, but let's confirm.

File: main.py
Comment (gretelliz) [reply main.py:216]: @viniarck, Good point, I hadn't realised. I have launched a [PR](https://github.com/kytos-ng/sdntrace_cp/pull/49)

File: main.py
Comment (gretelliz) [reply main.py:50]: Good point!

File: openapi.yml
Comment (gretelliz) [reply openapi.yml:7]: @viniarck, I have to modify the other endpoints too, right?

**Comment by gretelliz** [comment]
Ok,

@italovalcy and @jab1982 can you check this, please?

File: openapi.yml
Comment (viniarck) [reply openapi.yml]: @gretelliz, yes, putting a `v1` in the route before of the name of the resource like `/api/amlight/flow_stats/v1/flow/stats`.

**Review by viniarck — COMMENTED** [review]
@gretelliz great to have this enhancement here too, and fewer shared memory structures when it makes sense. 

The connection errors that you're seeing on tests seem to be related to `requests` not being completely mocked.

File: openapi.yml
Comment (viniarck) [reply openapi.yml]: We'll need this `v1` on all routes that you're stabilizing and making the v1 official.

**Comment by gretelliz** [comment]
@viniarck, you're welcome and thank you for your comments and suggestions.

**Review by viniarck — COMMENTED** [review]
@gretelliz, great to see that the test coverage has improved, the rest of coverage to achieve overall 85% let's leave it to issue #26. 

Overall looks good to me. The only things remaining for me to approve are:

1 -  [To have `sdntrace_cp` PR](https://github.com/kytos-ng/sdntrace_cp/pull/49) ready to land shortly after this one here is ready to get merged. 
2 - Double check with Italo the endpoint being removed, this one looks like the overall consensus is OK, including what was briefly discussed in our last team meeting. 
3 - [match_fields.py](https://github.com/kytos-ng/flow_stats/blob/master/match_fields.py) module doesn't seem to be used, I think we can nuke this to increase coverage a bit, since you're matching on stored flows (on sdntrace_cp). Could you double check? I have an impression that we shouldn't need to move these classes from this module to sdntrace_cp either. 

Thanks

File: main.py
Comment (gretelliz) [reply main.py:120]: I agree.

**Comment by gretelliz** [comment]
1. Now sdntrace_cp seems to be working fine.

2. Italo confirmed to me that he doesn't need the endpoints.

3. You are right, match_fields.py is not used and the coverage increased (80%) when it is removed.

**Comment by gretelliz** [comment]
Code cov in in 87% now. This is related to issue #26

**Review by viniarck — APPROVED** [review]
LGTM @gretelliz, it's great to have these improvements towards v1 stabilization and the extra test coverage.

**Comment by gretelliz** [comment]
@viniarck, thanks for your reviews and comments.

**Pull request merged**
